ENH: Use wheels for python packages

Eric89GXL opened the pull request:
I built some in a 14.04 x86_64 VM, let's see if 12.04 is happy. If this doesn't work, I might try using 12.04 x86_64, since that's more likely to be future compatible than 14.04 is to be backward compatible. If that doesn't work, we'll have to think harder about how to deal with library versions.

Eric89GXL commented:
Well, that bodes well that Travis is happy. On Monday I'll look into whether using this setup script on my 15.04 system actually works. It would be nice if it did!

BTW Travis seems to also have the "permission denied" error from #59. Not sure if there is a good workaround other than changing perms on `/var/log/nginx/error.log`.

Eric89GXL commented:
I built the packages on my 15.04 system, and didn't work on Travis due to a linking error (https://travis-ci.org/scitran/scitran/builds/76630002).

I then built the packages an a 12.04 x86_64 VM, and the packages worked on Travis, but not on my local system. So it looks like there is a system version dependence, which I don't find too surprising because the versions of lapack and atlas changed by a couple minor ticks.

Now I have built them both for 12.04 and 15.04, and added some checks to the script to make sure the right version is available. I also wrote a script that can be used on other systems (e.g., if a new Ubuntu version is needed) that will make the correct files. Does this approach seem okay?

An alternative would be to build both numpy and scipy *without* BLAS and LAPACK, which should work on any system with the basic C libraries (I think), but would be substantially slower. My vote would be to try using this optimized per-system method for a while, and if the multiple builds become problematic, start doing a general non-optimized build that will hopefully work on any x86_64 Ubuntu with the right GCC (and maybe python?) libs.

Eric89GXL commented:
Title updated, this works on my 15.04 locally as well as Travis 12.04. Ready for review/merge from my end.

Eric89GXL commented on scripts/0-setup.sh:
```diff
@@ -87,12 +78,23 @@ function EnsurePipPackages() {(
+           exit 1;
+	fi;
+
+	pip install --upgrade pip wheel setuptools
```
IMO upgrading pip here in the venv is a little bit nicer than doing `sudo pip install --upgrade pip` above

kofalt replied:
Agreed. As long as this check doesn't take too long (no reason to expect it would)

Eric89GXL commented on scripts/0-setup.sh:
```diff
@@ -87,12 +78,23 @@ function EnsurePipPackages() {(
+	pip install -q --no-index -f $url -r requirements/00_build_install.txt
+	pip install -q --no-index -f $url -r requirements/01_build_install.txt
+	pip install -q --no-index -f $url -r requirements/02_build_install.txt
+	pip install -q --no-index -f $url -r requirements/03_install.txt
```
previously an `egrep` was used to clean up these outputs, I have changed it to `-q` -- should I go back to `egrep`?

kofalt replied:
IMO yes, it'll be easier to use when we add fallback.

Eric89GXL commented on scripts/1-install.sh:
```diff
@@ -78,7 +78,8 @@ function EnsureNginx() {(
 	sudo mkdir -p /var/log/nginx/
-	sudo chmod 777 /var/log/nginx/
+	sudo touch /var/log/nginx/error.log
+	sudo chmod -R 777 /var/log/nginx/
```
this gets rid of the `permission denied` error that was still cropping up on Travis

Eric89GXL commented:
@gsfr I'll trigger a build here to see if the `zip` errors in `scitran` are now gone due to the updated code in `data`

Eric89GXL commented:
@gsfr zips now appear to work, but one of them has a hiccup:

https://travis-ci.org/scitran/scitran/builds/76683011#L589

In any case, this PR still appears to work.

gsfr commented:
The hiccup on ```no_json.zip``` is expected. It is an intentionally bad file to test *quarantining*.

Eric89GXL commented:
Alright, this one has been rebased (conflict in `0-setup.py` with recent commits to `master`) and is ready for review my end, assuming we do want to go with building wheels for each Ubuntu release. Travis is happy.

If it is ready for merge, then we can talk over email about where to host the files, because right now I'm using a link to our UW server.

kofalt commented:
Awesome. I hope to have a look at this within the next few days.

Eric89GXL commented:
Rebased again, new packages built and pushed, and the CIs are happy.

kofalt commented on scripts/2-run.sh:
```diff
@@ -104,6 +104,7 @@ function BootStrapUsers() {(
 		StopReflex
 		exit $result;
 	fi
+	set -e
```
Smart, but I think unnecessary - I throw subshells around like there's no tomorrow to avoid that kind of problem. I'm pretty sure the environment dies when the subshell `()` closes on the next line.

Eric89GXL replied:
It's more of a cleanup/consistency thing because `set +e` is done above. It seems cleaner to have the matching pair, no?

kofalt replied:
*shrug* Sure, we can keep it.

kofalt commented on scripts/2-run.sh:
```diff
@@ -132,6 +133,7 @@ function BootStrapData() {(
 		StopReflex
 		exit $result;
 	fi
+	set -e
```
Same here

kofalt commented:
Questions:

1. What does the `requests[security]` syntax in requirements do?
1. Why is requests specified in multiple requirements files?
1. What purpose does repeating deps in both `03_build.txt` and `03_install.txt` serve?
1. Is the purpose of spreading out so many requirements files to order dep management?
1. Why remove the `uuidgen` install snippet? IMO save that for the source-infra discussion.
1. Similarly, why remove the `ca-certificates` install snippet?

Comments:

1. Let's move both the `arch` and `release` detection to `DetectPlatform` in `0-setup.sh`.
1. Unfortunately `lsb_release -sr` won't do what you're hoping for Ubuntu derivatives, for example Linux Mint. The good news is that all modern Ubuntu releases have a file `/etc/os-release` ([example](http://hastebin.com/uwezezusiz.ini)) with this information. So in based on `platform` in ``DetectPlatform`, I would instead use `cat /etc/os-release | grep VERSION_ID | cut -f 2 -d '"'` or some cleverer equivalent.

kofalt commented:
Overall this looks outstanding :+1: 

Let's tackle everything in the above review first, then move the discussion to how we can build & host these wheels in-house. My inclination would be to ferret away N vagrantfiles that run your script, then host the results on our cloud storage. I'd also like the script to fall back on compiling these packages for platforms that we haven't build wheels for, since running on OSX natively is desired (ref #35).

Eric89GXL commented:
> What does the requests[security] syntax in requirements do?

It was meant to install the `requests` package with the `security` option to squash the warnings about no secure SSL context. It didn't quite work, but should be better than not having the `security` option at all.

> Why is requests specified in multiple requirements files?

Ahh, it should only be in `00` not `01`.

> What purpose does repeating deps in both 03_build.txt and 03_install.txt serve?

They are slightly different -- take a look. If you are building from git source, then you need to separate. If we only used one file, then `pip` doesn't know if something has changed on `GitHub` and so *always* rebuilds and reinstalls. By building, checking what the resulting version number is, and installing that resulting version number, you end up avoiding the recompile/reinstall for every check. That's what this comment is meant to allude to:

https://github.com/scitran/scitran/pull/61/files#diff-f90487b4180c3b354b3486096086bcbcR33

I can make it clearer if necessary.

> Is the purpose of spreading out so many requirements files to order dep management?

Yep. Otherwise e.g. `mne` can't install because its install script requires `numpy`.

> Why remove the uuidgen install snippet? IMO save that for the source-infra discussion.

I meant to move the `uuid-runtime` install to the `EnsurePackages` call, I'll move it.

> Similarly, why remove the ca-certificates install snippet?

It is still there, it has just been moved. This made all the `sudo apt-get` calls be made in one place. Eventually with infra this can then easily be factored out.

I'll take care of the other two comments, as well.

Should we do the fallback compilation in a separate PR? It probably makes sense to refactor the wheel-building script so that we can run that in case the architecture is not supported.

Eric89GXL commented:
Correction: `uuid-runtime` is already in there, too, it was also moved

kofalt commented:
> By building, checking what the resulting version number is, and installing that resulting version number, you end up avoiding the recompile/reinstall for every check.

Gotcha. I didn't realize the two `03` files were separately for building wheels versus install workflow. I think I'll add a `readme.md` to the new `requirements` folder, and in the install + build cases use a bash iterator over a glob. I think that'll clear up the distinction.

> Should we do the fallback compilation in a separate PR?

Yes, I think that's reasonable. We can just roll that into the requirements for #35.

---

Would you be open to adding me as a collaborator to your fork? 

After you're satisfied with the pending changes, let me know and I'll add some quick tune-up commits (such as the doc + glob change mentioned above, and likely some vagrant scripting).

Alternatively, we could re-target this PR to a dev branch that I'll add onto after it lands.

Eric89GXL commented:
> in the install + build cases use a bash iterator over a glob. I think that'll clear up the distinction.

Unfortunately we can't iterate over the builds, then iterate over the installs if that's what you have in mind. Each step needs to be built, then installed, to satisfy dependencies for the next step so that won't work. If you have in mind to use globbing while preserving this order, then yeah it's probably better if you do it :) I guess you could loop over `01, 02, 03` (or auto-detect these) and use globbing to find the right `build` and `install` text files...?

I've added you as a repo collaborator. Note you can also open a PR into my branch, the only difference between that and opening a PR from a given branch into `upstream:master` (std. behavior) is that you'd select a different fork and branch (`Eric89GXL:wheels`) on the GitHub interface. But if it's easier/cleaner to push directly I'm fine with that, too.

Rebased, and comments addressed. Feel free to push additional changes.

Eric89GXL commented:
(only the `FIX: Better comments` commit above is new since your last review)

Eric89GXL commented:
Looks like Travis is unhappy now for some reason. You going to work on this now? If so, I'll hold off on trying to fix Travis. If not, I can tackle it in about an hour.

gsfr commented:
>> What does the requests[security] syntax in requirements do?

> It was meant to install the requests package with the security option to squash the warnings about no secure SSL context. It didn't quite work, but should be better than not having the security option at all.

Is this the "insecure platform" warning? If so, we shouldn't be turning that off. We should rather be on a version of python that's actually secure. I'm on 2.7.10, which works without any warnings.

Eric89GXL commented:
I'm pretty sure the `requests[security]` is supposed to make it secure. Something about monkey-patching to get it up to snuff. So "squash the warnings" is probably not the right phrase, but rather "improve security so that no warnings are necessary".

gsfr commented:
Ok. But why not just go with the latest (and secure) version of Python and no monkey patching?

kofalt commented:
I think ubuntu-packaged python triggers the warning. I've kinda assumed the python maintainers and ubuntu maintainers disagree on something & I ignore the warning.

kofalt commented:
@Eric89GXL go ahead and fix CI; I don't know why pip can't find those packages. For now I think I'll just add commits to your branch (once you're satisfied) but it'll likely be at least tomorrow.

Eric89GXL commented:
@gsfr right now we use the packaged version as @kofalt says, so we'd need to build our own version of Python to avoid the warning, unless we can get a newer version from a PPA or something. A bit overkill IMO, especially if we can get `pip` installing to take care of it for us instead. But this also isn't a problem on 15.04 (or probably even 14.04), it's only a problem because Travis runs 12.04 so probably not worth thinking about too hard.

I'll see if I can get the CIs happy again.

gsfr commented:
Fair enough, if it works. Still an issue on 14.04, though, which ships with 2.7.6.

kofalt commented:
@gsfr IMO file a bug on the requests security problem, as it's something that pre-exists this PR. I am open with using either `requests[security]` or just `requests` based on what you two are comfortable with.

If you're thinking that we should bundle a python distribution, that would be a second bug to file... I took a [quick look](https://www.python.org/downloads/release/python-2710/) and did not see any tarball-binary distributions. They might be hosted elsewhere.

Eric89GXL commented:
I just ran around in circles for a while to see that:
```
/etc/os-release: No such file or directory
```
@kofalt looks like that won't work for detecting the version unfortunately. We might need to triage.

Eric89GXL commented:
Okay, CIs are happy. I changed the Bash code to use loops. @kofalt see if it's what you had in mind, if not, feel free to push to my branch.

kofalt commented:
Looks good, using the os-release file if it exists sounds reasonable.

One small problem:

```
scripts/0-setup.sh: line 75: [: missing `]'
scripts/0-setup.sh: line 75: 1: command not found
```

Looks like it takes issue with the `if [ -z "$release" || ...` line.

kofalt commented:
I will add commits soon (tomorrow? friday?) that:

* Add some vagrants for the most-recent ubuntu releases
* A quick script that goes through each of those
* Change the download URL

Then:

* Upload our new wheels
* Test it out, wait for CI
* Merge :)

Eric89GXL commented:
Fixed the bash conditional problem.

You should be able to pull the wheels you need from https://lester.ilabs.uw.edu/files/wheelhouse to upload wherever you need. Or you can email me offline for how to get access to your server and I can do it. That might be less time consuming because I can use e.g. ssh / rsync to pull and push all the files easily.

Eric89GXL commented:
ping @kofalt just a friendly reminder about this if you have time

kofalt commented:
Thanks for the bump. I've made a few additions:

* Vagrant now builds wheels for a few releases
* Hosting is now on our Google bucket
* New scripting for building + installing wheels
* Simplify requirements processing

The new `requirements/Vagrantfile` defines a set of VMs for various ubuntu releases.
One runs the convenience script `requirements/create-wheels.sh` to spin each up in turn.

Each vagrant will use our pre-existing `scripts/vagrant-create.sh`, then the new `scripts/vagrant-create-wheels.sh` to install various packages. While setting this up, I found the instructions required a few more packages (liblapack-dev, libatlas-base-dev, gfortran).

The result files land in `requirements/out`. We're hosting them on our Google storage bucket.
Unfortunately, that is not compatible with pip's URL-discovery feature, so I have a temporary hackaround.

If for any reason downloads fail (availability, unsupported release) installation happens manually as before.

I didn't run into idempotency problems with URL installs, so I was able to re-simplify requirement file order.
Logic is straightforward: `*basic*` for zero-compile deps, `*wheels*` for packages that may or may not be pre-compiled, and `*source*` for github sources.

kofalt commented:
Or, to summarize:

1. We're now hosting 12.04, 14.04, and 15.04 linux x64 builds of numpy, scipy, and mne
1. These are optimistically and (for now) hackishly installed if available
1. Optimistic download failures are ignored, and any failed packages are built manually
1. Vagrant configuration was written & used to generate these; will be easy to add more in the future.

You can view the diff of just my changes [here](https://github.com/scitran/scitran/compare/4c9a732...Eric89GXL:wheels).
CI is lit, so I'm prepared to merge this pending your approval. @gsfr gave it a quick sanity-check.

Eric89GXL commented:
Great! I'll pull and test locally and check out the changes, but your changes sound reasonable to me.

Eric89GXL commented:
@kofalt works great over here, +1 for merge. Thanks for all the changes, they make it cleaner and more robust.

Eric89GXL commented:
After this gets merged, does it make sense to tackle #2 or #44?

kofalt commented:
Solid. This is great to land, I greatly appreciate the help learning wheels :clap: 

#44, definitely. I had another look at the diff, it still looks reasonable. I'll rebase after merging this.

I want to make progress on #2 also - I have some thoughts I need to compose + post there. Most likely #2 will hold until after #44. Internally we've been preparing to cut a release towards the end of this month, and after that would be the most opportune time for refactoring.

gsfr commented:
I agree with @kofalt. #44 and #35 are the next steps.

Re #2: I have been making some progress in planning out how to evolve the various components (```reaper```, ```api```, ```data```) to decouple them. Customer demand needs us to be able to reap and manage data of any type, without prior knowledge. Metadata extraction thus becomes an asynchronous *engine* task. Adding support for a new data type should only require updating the ```data``` component, unless ```reaper``` needs to learn a new networking protocol.

To that end, I am taking up work on scitran/reaper#1 and scitran/reaper#5. @rentzso is working on matching ```api``` changes and @lmperry will update our *dmc2nii* container to additionally provide DICOM metadata.

What we need now is a full specification of the *engine* interface for *SciTran Apps*, aka *Flywheel Gears*.
